srml/support/procedural/src/construct_runtime/mod.rs

```diff
@@ -0,0 +1,395 @@
+// Copyright 2019 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+mod parse;
+
+use parse::{ModuleDeclaration, ModulePart, RuntimeDefinition, WhereSection};
+use proc_macro::TokenStream;
+use proc_macro2::{Span, TokenStream as TokenStream2};
+use quote::quote;
+use srml_support_procedural_tools::syn_ext as ext;
+use srml_support_procedural_tools::{generate_crate_access, generate_hidden_includes};
+use syn::Ident;
+
+// try macro but returning tokenized error
+macro_rules! try_tok(( $expre : expr ) => {
+	match $expre {
+		Ok(r) => r,
+		Err (err) => {
+			return err.to_compile_error().into()
+		}
+	}
+});
+
+pub fn construct_runtime(input: TokenStream) -> TokenStream {
+	let definition = syn::parse_macro_input!(input as RuntimeDefinition);
+	let RuntimeDefinition {
+		name,
+		where_section: WhereSection {
+			block,
+			node_block,
+			unchecked_extrinsic,
+			..
+		},
+		modules: ext::Braces {
+			content: ext::Punctuated { inner: modules, .. },
+			token: modules_token,
+		},
+		..
+	} = definition;
+
+	// Assert we have system module declared
+	let system_module = match find_system_module(modules.iter()) {
+		Some(sm) => sm,
+		None => {
+			return syn::Error::new(
```

Comment:
I think this is really bad style. This `construct_runtime` should return a `Result` and this `Result` should be handled in the calling method.